pennylane/beta/tapes/tape.py

```diff
@@ -0,0 +1,776 @@
+
+        **Example**
+
+        .. code-block:: python
+
+            from pennylane.beta.tapes import QuantumTape
+            from pennylane.beta.queuing import expval, var, sample, probs
+
+            with QuantumTape() as tape:
+                qml.RX(0.432, wires=0)
+                qml.RY(0.543, wires=0)
+                qml.CNOT(wires=[0, 'a'])
+                qml.RX(0.133, wires='a')
+                expval(qml.PauliZ(wires=[0]))
+
+        >>> tape.operations
+        [RX(0.432, wires=[0]), RY(0.543, wires=[0]), CNOT(wires=[0, 'a']), RX(0.133, wires=['a'])]
+        """
+        return self._prep + self._ops
+
+    @property
+    def observables(self):
+        """Returns the observables on the quantum tape.
+
+        Returns:
+            list[.Observable]: list of recorded quantum operations
+
+        **Example**
+
+        .. code-block:: python
+
+            from pennylane.beta.tapes import QuantumTape
+            from pennylane.beta.queuing import expval, var, sample, probs
+
+            with QuantumTape() as tape:
+                qml.RX(0.432, wires=0)
+                qml.RY(0.543, wires=0)
+                qml.CNOT(wires=[0, 'a'])
+                qml.RX(0.133, wires='a')
+                expval(qml.PauliZ(wires=[0]))
+
+        >>> tape.operations
+        [expval(PauliZ(wires=[0]))]
+        """
+        return [m[1] for m in self._obs]
+
+    @property
+    def measurements(self):
+        """Returns the measurements on the quantum tape.
+
+        Returns:
+            list[.MeasurementProcess]: list of recorded measurement processess
+
+        **Example**
+
+        .. code-block:: python
+
+            from pennylane.beta.tapes import QuantumTape
+            from pennylane.beta.queuing import expval, var, sample, probs
+
+            with QuantumTape() as tape:
+                qml.RX(0.432, wires=0)
+                qml.RY(0.543, wires=0)
+                qml.CNOT(wires=[0, 'a'])
+                qml.RX(0.133, wires='a')
+                expval(qml.PauliZ(wires=[0]))
+
+        >>> tape.measurements
+        [<pennylane.beta.queuing.measure.MeasurementProcess object at 0x7f10b2150c10>]
+        """
+        return [m[0] for m in self._obs]
+
+    @property
+    def num_params(self):
+        """Returns the number of trainable parameters on the quantum tape."""
+        return len(self.trainable_params)
+
+    @property
+    def output_dim(self):
+        """The (estimated) output dimension of the quantum tape."""
+        return self._output_dim
+
+    @property
+    def diagonalizing_gates(self):
+        """Returns the gates that diagonalize the measured wires such that they
+        are in the eigenbasis of the circuit observables.
+
+        Returns:
+            List[~.Operation]: the operations that diagonalize the observables
+        """
+        rotation_gates = []
+
+        for observable in self.observables:
+            rotation_gates.extend(observable.diagonalizing_gates())
```

Comment:
I remember discussing at one point making it possible for observables to be evaluated on the same wires. I guess this method is incompatible with that, but maybe that's a question for the future.

Reply:
Yep. If we want to measure different observables on the same wires, we have two options:

* Simulators only: simply calculate <psi|A|psi>and <psi|B|psi>

* Hardware only: A and B must commute. Determine the rotation that rotates the circuit into the shared eigenbasis.